### PR TITLE
UCT/API: add flags to active message callbacks

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe @shamisp 
- api change
- add am callback flag that allows callback invocation from
async event. (ex progress thread)

### CODE REVIEW

File: src/ucp/wireup/wireup.c [hunk 1]
```diff
@@ -869,7 +868,7 @@ static ucs_status_t ucp_select_transport(ucp_worker_h worker, ucp_address_t *add
 ucs_status_t ucp_wireup_set_am_handlers(ucp_worker_h worker, uct_iface_h iface)
 {
     return uct_iface_set_am_handler(iface, UCP_AM_ID_WIREUP,
-                                    ucp_wireup_msg_handler, worker);
+                                    ucp_wireup_msg_handler, worker, UCT_AM_CB_FLAG_MT_SAFE);
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c]: wireup does not have to be thread safe

File: src/ucp/core/ucp_worker.c [hunk 2]
```diff
@@ -32,7 +32,7 @@ static ucs_status_t ucp_worker_set_am_handlers(ucp_worker_h worker,
     for (am_id = 0; am_id < UCP_AM_ID_LAST; ++am_id) {
         if (context->config.features & ucp_am_handlers[am_id].features) {
             status = uct_iface_set_am_handler(iface, am_id, ucp_am_handlers[am_id].cb,
-                                              worker);
+                                              worker, 0);
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: better to add that flag to ucp_am_handlers

File: src/ucp/wireup/wireup.c [hunk 3]
```diff
@@ -869,7 +868,7 @@ static ucs_status_t ucp_select_transport(ucp_worker_h worker, ucp_address_t *add
 ucs_status_t ucp_wireup_set_am_handlers(ucp_worker_h worker, uct_iface_h iface)
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c]: need to remove this function

File: src/uct/api/uct.h [hunk 4]
```diff
@@ -636,9 +636,15 @@ void uct_iface_mem_free(const uct_allocated_memory_t *mem);
  * @param [in]  id       Active message id. Must be 0..UCT_AM_ID_MAX-1.
  * @param [in]  cb       Active message callback. NULL to clear.
  * @param [in]  arg      Active message argument.
+ * @param [in]  flags    Required capability flags
```
Comment (yosefe) [inline src/uct/api/uct.h]: IMHo it's "Active message callback capabilities"

File: src/uct/api/uct.h [hunk 5]
```diff
@@ -636,9 +636,15 @@ void uct_iface_mem_free(const uct_allocated_memory_t *mem);
...
  */
+enum {
+    UCT_AM_CB_FLAG_MT_SAFE = UCS_BIT(1), /* callback can be safely called
```
Comment (yosefe) [inline src/uct/api/uct.h]: `UCT_AM_FLAG_THREAD_SAGE` - "callback can be called from any thread (by default it would be called only a thread which explicitly calls uct_iface_progress"

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1382/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/771/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
looks good, IMHO we should come up with better name for the flag, now it's a bit confising

File: test/gtest/uct/test_ib.cc [hunk 6]
```diff
@@ -163,7 +163,7 @@ UCS_TEST_P(test_uct_ib, non_default_pkey, "IB_PKEY=0x2")
     recv_buffer->length = 0; /* Initialize length to 0 */
 
     /* set a callback for the uct to invoke for receiving the data */
-    uct_iface_set_am_handler(m_e2->iface(), 0, ib_am_handler , recv_buffer);
+    uct_iface_set_am_handler(m_e2->iface(), 0, ib_am_handler , recv_buffer, 0);
```
Comment (shamisp) [inline test/gtest/uct/test_ib.cc]: shell we define a separate NONE flag

File: test/gtest/uct/test_p2p_am.cc [hunk 7]
```diff
@@ -150,7 +150,7 @@ class uct_p2p_am_test : public uct_p2p_test
             short_progress_loop();
         }
 
-        status = uct_iface_set_am_handler(receiver().iface(), AM_ID, NULL, NULL);
+        status = uct_iface_set_am_handler(receiver().iface(), AM_ID, NULL, NULL, 0);
```
Comment (shamisp) [inline test/gtest/uct/test_p2p_am.cc]: I don't see any tests with the flag set to 1 ?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1385/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/774/ for details (Mellanox internal link).

File: test/gtest/uct/test_p2p_am.cc (hunk 7 above)
Comment (yosefe) [inline test/gtest/uct/test_p2p_am.cc]: Yep. @alex-mikheev maybe we could add it as another dimension to uct_p2p_am test?

File: test/gtest/uct/test_ib.cc (hunk 6 above)
Comment (shamisp) [inline test/gtest/uct/test_ib.cc]: I would add UCT_AM_CB_FLAG_NONE = 0

Comment (yosefe) [inline test/gtest/uct/test_ib.cc]: UCT_AM_CB_FLAG_DEFAULT

Comment (shamisp) [inline test/gtest/uct/test_ib.cc]: +1

**Comment by MattBBaker** [comment]
autobot build

File: test/gtest/uct/test_p2p_am.cc (hunk 7 above)
Comment (alex-mikheev) [inline test/gtest/uct/test_p2p_am.cc]: i plan to do it in another PR together with UD changes

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1393/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/781/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
:+1:

**Comment by shamisp** [comment]
:+1:
